base/setup/lib/utils/partlist.c

```diff
@@ -2878,124 +2933,67 @@ CreatePartition(
-        return Status;
-    }
+        /* Check whether the partition is valid and was mounted by the system */
+        ASSERT(Volume->PartEntry == PartEntry);
+        ASSERT(PartEntry->IsPartitioned && PartEntry->PartitionNumber != 0);
+        ASSERT(PartEntry->PartitionType != PARTITION_ENTRY_UNUSED);
+        ASSERT(!IsContainerPartition(PartEntry->PartitionType));
 
-    /* Lock the volume */
-    LockStatus = NtFsControlFile(PartitionHandle,
-                                 NULL,
-                                 NULL,
-                                 NULL,
-                                 &IoStatusBlock,
-                                 FSCTL_LOCK_VOLUME,
-                                 NULL,
-                                 0,
-                                 NULL,
-                                 0);
-    if (!NT_SUCCESS(LockStatus))
-    {
-        DPRINT1("WARNING: Failed to lock volume! Operations may fail! (Status 0x%lx)\n", LockStatus);
-    }
+        /* Dismount the basic volume: unlink the volume
+         * from the list of existing ones */
+        PartEntry->Volume = NULL;
+        Volume->PartEntry = NULL;
+        RemoveEntryList(&Volume->ListEntry);
 
-    /* Dismount the volume */
-    Status = NtFsControlFile(PartitionHandle,
-                             NULL,
-                             NULL,
-                             NULL,
-                             &IoStatusBlock,
-                             FSCTL_DISMOUNT_VOLUME,
-                             NULL,
-                             0,
-                             NULL,
-                             0);
-    if (!NT_SUCCESS(Status))
-    {
-        DPRINT1("Failed to unmount volume (Status 0x%lx)\n", Status);
+        // NOTE: If FormatState == Unformatted but *FileSystem != 0 this means
+        // it has been usually mounted with RawFS and thus needs to be dismounted.
+        if (Volume->FormatState == UnknownFormat)
+        {
+            RtlFreeHeap(ProcessHeap, 0, Volume);
+            Volume = NULL;
+        }
     }
-
-    /* Unlock the volume */
-    LockStatus = NtFsControlFile(PartitionHandle,
-                                 NULL,
-                                 NULL,
-                                 NULL,
-                                 &IoStatusBlock,
-                                 FSCTL_UNLOCK_VOLUME,
-                                 NULL,
-                                 0,
-                                 NULL,
-                                 0);
-    if (!NT_SUCCESS(LockStatus))
+    if (!Volume)
     {
-        DPRINT1("Failed to unlock volume (Status 0x%lx)\n", LockStatus);
+        /* The partition is not mounted, so just return success */
+        return STATUS_SUCCESS;
     }
 
-    /* Close the volume */
-    NtClose(PartitionHandle);
-
-    return Status;
+    /* Dismount the basic volume: link the volume into the list of volumes to unmount */
+    // return DismountVolume(&Volume->Info);
```

Comment:
```suggestion
    // return DismountVolume(&Volume->Info); //FIXME: does not work yet
```

Reply:
This works actually, except that this is disabled until the partlist.c code supports both delayed-pending (as it currently done), and immediate operations, via some sort of mode flag. In immediate-operatins case, it is here that the DismountVolume call would be done.